src/main/java/jcifs/netbios/NameServiceClientImpl.java

```diff
@@ -0,0 +1,1072 @@
+                return true;
+            }
+        }
+        return false;
+    }
+
+
+    protected InetAddress switchWINS () {
+        this.nbnsIndex = ( this.nbnsIndex + 1 ) < this.transportContext.getConfig().getWinsServers().length ? this.nbnsIndex + 1 : 0;
+        return this.transportContext.getConfig().getWinsServers().length == 0 ? null
+                : this.transportContext.getConfig().getWinsServers()[ this.nbnsIndex ];
+    }
+
+    static class Sem {
+
+        Sem ( int count ) {
+            this.count = count;
+        }
+
+        int count;
+    }
+
+    static class QueryThread extends Thread {
+
+        private Sem sem;
+        private String host, scope;
+        private int type;
+        private NetbiosAddress ans = null;
+        private InetAddress svr;
+        private UnknownHostException uhe;
+        private CIFSContext tc;
+
+
+        QueryThread ( Sem sem, String host, int type, String scope, InetAddress svr, CIFSContext tc ) {
+            super("JCIFS-QueryThread: " + host);
+            this.sem = sem;
+            this.host = host;
+            this.type = type;
+            this.scope = scope;
+            this.svr = svr;
+            this.tc = tc;
+        }
+
+
+        @Override
+        public void run () {
+            try {
+                this.ans = this.tc.getNameServiceClient().getNbtByName(this.host, this.type, this.scope, this.svr);
+            }
+            catch ( UnknownHostException ex ) {
+                this.uhe = ex;
+            }
+            catch ( Exception ex ) {
+                this.uhe = new UnknownHostException(ex.getMessage());
+            }
+            finally {
+                synchronized ( this.sem ) {
+                    this.sem.count--;
+                    this.sem.notify();
+                }
+            }
+        }
+
+
+        /**
+         * @return the ans
+         */
+        public NetbiosAddress getAnswer () {
+            return this.ans;
+        }
+
+
+        /**
+         * @return the uhe
+         */
+        public UnknownHostException getException () {
+            return this.uhe;
+        }
+
+    }
+
+
+    NetbiosAddress lookupServerOrWorkgroup ( String name, InetAddress svr ) throws UnknownHostException {
+        Sem sem = new Sem(2);
+        int type = isWINS(svr) ? 0x1b : 0x1d;
+
+        QueryThread q1x = new QueryThread(sem, name, type, null, svr, this.transportContext);
+        QueryThread q20 = new QueryThread(sem, name, 0x20, null, svr, this.transportContext);
+        q1x.setDaemon(true);
+        q20.setDaemon(true);
+        try {
+            synchronized ( sem ) {
+                q1x.start();
+                q20.start();
+
+                while ( sem.count > 0 && q1x.getAnswer() == null && q20.getAnswer() == null ) {
+                    sem.wait();
+                }
+            }
+        }
+        catch ( InterruptedException ie ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)